src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//==== ONNXToTosaLegalizeUtils.hpp - ONNX dialects to TOSA lowering Utils-===//
+//
+// Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file contains common utils shared by the functions performing the
+// lowering to the TOSA dialect. It is also used by TensorFlow and torch-mlir.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef ONNXMLIR_CONVERSION_ONNXTOTOSA_TOSALEGALIZEUTILS_H
+#define ONNXMLIR_CONVERSION_ONNXTOTOSA_TOSALEGALIZEUTILS_H
+
+#include "mlir/Dialect/Quant/QuantTypes.h"        // from @llvm-project
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"         // from @llvm-project
+#include "mlir/Dialect/Tosa/Utils/ShapeUtils.h"   // from @llvm-project
+#include "mlir/IR/BuiltinAttributes.h"            // from @llvm-project
+#include "mlir/IR/BuiltinTypes.h"                 // from @llvm-project
+#include "mlir/IR/PatternMatch.h"                 // from @llvm-project
+#include "mlir/Interfaces/InferTypeOpInterface.h" // from @llvm-project
+#include "mlir/Support/LLVM.h"                    // from @llvm-project
+
+namespace onnx_mlir {
+namespace tosa {
+
+// Creates a TOSA operation and performs shape inference on the individual
+// op. This allows shape inference during the framework to TOSA lowering.
+template <typename TosaOp, typename... Args>
+TosaOp CreateOpAndInfer(mlir::PatternRewriter &rewriter, mlir::Location loc,
+    mlir::Type result_ty, Args &&... args) {
+  auto op = rewriter.create<TosaOp>(loc, result_ty, args...);
+
+  mlir::InferShapedTypeOpInterface shapeInterface =
+      llvm::dyn_cast<mlir::InferShapedTypeOpInterface>(op.getOperation());
+  if (!shapeInterface)
+    return op;
+
+  llvm::SmallVector<mlir::ShapedTypeComponents> returnedShapes;
+  if (shapeInterface
+          .inferReturnTypeComponents(op.getContext(), op.getLoc(),
+              op->getOperands(), op->getAttrDictionary(), op->getRegions(),
+              returnedShapes)
+          .failed())
+    return op;
+
+  // We need to use the element type of the existing result type to generate
+  // the new result shaped type. This is because rescale can include a cast to
+  // different bit-width types and does not have a TypeAttr to define the
+  // target type.
```

Comment:
We have a function [updateType](https://github.com/onnx/onnx-mlir/blob/main/src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp#L149)  doing the same job. E.g, `updateType(result, predictedShape, elementType)` will update the type of `result` by using `predictedShape` whenever feasible. Please consider to use it.

Reply:
While you are right, we would like to keep this as is to align as much as possible to [tf-to-tosa](https://github.com/tensorflow/tensorflow/blob/0cf4668f0c6e88ebc5cad993e4956199af2b23de/tensorflow/compiler/mlir/tosa/transforms/legalize_utils.h#L144-L198) and [torch-to-tosa](https://github.com/llvm/torch-mlir/blob/af9e8a5e63e2ca8d498d781d9d56f2c2f8362146/include/torch-mlir/Conversion/TorchToTosa/TosaLegalizeUtils.h#L59-L110).

Reply:
After talking with @ljfitz who had a conversation with @eric-k256, we decided to adopt `updateType` and Eric thinks of using this in TF, too.

Reply:
Cool. Thanks for using `updateType`! I hope you find it convenient.